[SPARK-24143] filter empty blocks when convert mapstatus to (blockId, size) pair

jinxing64 opened the pull request:
## What changes were proposed in this pull request?

In current code(`MapOutputTracker.convertMapStatuses`), mapstatus are converted to (blockId, size) pair for all blocks – no matter the block is empty or not, which result in OOM when there are lots of consecutive empty blocks, especially when adaptive execution is enabled.

(blockId, size) pair is only used in `ShuffleBlockFetcherIterator` to control shuffle-read and only non-empty block request is sent. Can we just filter out the empty blocks in MapOutputTracker.convertMapStatuses and save memory?


## How was this patch tested?

not added yet.

Commit:
[SPARK-24143] filter empty blocks when convert mapstatus to (blockId, size) pair.

jinxing64 commented:
@squito @cloud-fan @jiangxb1987 

Do you think this make sense?

SparkQA commented:
**[Test build #90032 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90032/testReport)** for PR 21212 at commit [`5211ebd`](https://github.com/apache/spark/commit/5211ebd5cc4d3de023752b8ab8168d7bda18aa83).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2795/
Test PASSed.

SparkQA commented:
**[Test build #90032 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90032/testReport)** for PR 21212 at commit [`5211ebd`](https://github.com/apache/spark/commit/5211ebd5cc4d3de023752b8ab8168d7bda18aa83).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90032/
Test FAILed.

SparkQA commented:
**[Test build #90041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90041/testReport)** for PR 21212 at commit [`5211ebd`](https://github.com/apache/spark/commit/5211ebd5cc4d3de023752b8ab8168d7bda18aa83).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2804/
Test PASSed.

jiangxb1987 commented:
How much memory did the converted pairs consume? If the empty blocks should be a issue can we just clean up the empty blocks?

cloud-fan commented:
I think it's reasonable to filter out empty shuffle blocks, shall we also optimize the space usage for `MapStatus`? then we can also reduce network traffic.

SparkQA commented:
**[Test build #90041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90041/testReport)** for PR 21212 at commit [`5211ebd`](https://github.com/apache/spark/commit/5211ebd5cc4d3de023752b8ab8168d7bda18aa83).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90041/
Test FAILed.

squito commented:
This makes sense to me.  You should update the comment on `MapOutputTracker.getMapSizesByExecutorId` to mention that it excludes the zero-sized blocks, and also remove the filter in `ShuffleBlockFetcherIterator` and fix up the related comments there to mention where the filtering is happening.

curious, have you observed this using a lot of memory from a heapdump or anything?  I think its a pretty reasonable change either way, I'd just like to know.

A couple more memory improvement you could throw in here as well

1. [`statuses.zipWithIndex`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/MapOutputTracker.scala#L863) will create a copy of the array, that is pointless.  you could either switch to `statuses.iterator.zipWithIndex` or just manually iterate over the indices.

2. the final [`splitsByAddress.toSeq`]( https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/MapOutputTracker.scala#L876) creates a copy, you could change types so that is not needed.  It could return a map, or if we really want to make sure its not modified, I think you could have it return an `Iterator` instead of a Seq so that you could just call `splitsByAddress.iterator`.

3. I wonder if using an ArrayBuffer is the smartest, maybe it'll have a ton of wasted space from the way it grows.  Its also only iterated over, we don't need fast random access.  perhaps a `List` or `Vector` might be better.  But this one I'm less sure about.

to be clear, this isn't changing network traffic at all -- this is after the executor has received the MapStatus, and its just doing some datastructure manipulation internal to the executor.

Commit:
improve memory usage and refine doc.

SparkQA commented:
**[Test build #90127 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90127/testReport)** for PR 21212 at commit [`2997a45`](https://github.com/apache/spark/commit/2997a45e8587115eebd66b9974b4a8dc56d821b1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2867/
Test PASSed.

jinxing64 commented:
@squito @cloud-fan @jiangxb1987
Thanks a lot for review.

> shall we also optimize the space usage for MapStatus

@cloud-fan do you mean optimize space usage for MapStatus when there are lots of consecutive empty-blocks ?

Rashid, thanks a lot for your comments. I refined the doc and did some more improvement for memory usage. Please take another look.
>curious, have you observed this using a lot of memory from a heapdump or anything?

Yes, it's from heapdump. We enabled adaptive execution and found this issue.

> I wonder if using an ArrayBuffer is the smartest, maybe it'll have a ton of wasted space from the way it grows.

I'm not sure about this either. So I just keep the ArrayBuffer here.

squito commented:
If you have a heap dump, there are tools that can check for wasted space in ArrayBuffer.  Eg. [jxray](http://www.jxray.com/) or [YourKit Memory Inspections](https://www.yourkit.com/docs/java/help/inspections_mem.jsp#sparse_arrays) perhaps.  if you can share the heapdump, we could analyze, or perhaps you could try it yourself if you can't share.

squito commented:
can you add a test in MapOutputTrackerSuite and update the pr description to include all the changes?  but overall looks good.

SparkQA commented:
**[Test build #90127 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90127/testReport)** for PR 21212 at commit [`2997a45`](https://github.com/apache/spark/commit/2997a45e8587115eebd66b9974b4a8dc56d821b1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90127/
Test PASSed.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -267,28 +269,28 @@ final class ShuffleBlockFetcherIterator(
       if (address.executorId == blockManager.blockManagerId.executorId) {
-        // Filter out zero-sized blocks
-        localBlocks ++= blockInfos.filter(_._2 != 0).map(_._1)
+        blockInfos.find(_._2 < 0) match {
```
shall we use `_._2 < 0` to make sure the 0-size blocks are filtered?

jinxing64 replied:
Yes, I added check for _._2 ==0

cloud-fan commented:
> do you mean optimize space usage for MapStatus when there are lots of consecutive empty-blocks ?

Yea, something like doing an RLE for the size array in `CompressedMapStatus`. But this can be done as a followup.

Commit:
ArrayBuffer->ListBuffer, unit test

jinxing64 commented:
@squito 
I ananlyzed "YourKit Memory Inspections" to analyze the heap, but didn't find many duplicate objects in ArrayBuffer. I guess your concern is ArrayBuffer will do lots of copy as size of elements grows, and we don't need fast random access in `ShuffleBlockFetcherIterator`. I changed it to be ListBuffer.
I also added a test in MapOutputTrackerSuite. Please take a look

SparkQA commented:
**[Test build #90243 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90243/testReport)** for PR 21212 at commit [`26442c1`](https://github.com/apache/spark/commit/26442c12619ee59bd98f70256870f83146934f37).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90243/
Test FAILed.

Ngone51 commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -267,28 +269,30 @@ final class ShuffleBlockFetcherIterator(
+          case Some((blockId, size)) if size < 0 =>
+            throw new BlockException(blockId, "Negative block size " + size)
+          case Some((blockId, size)) if size == 0 =>
+            throw new BlockException(blockId, "Zero-sized blocks should be excluded.")
```
Is it necessary to throw exception here? If so, shall we also throw exception when detect 0-sized **remote** block rather than skip it silently?

jinxing64 replied:
Added another check for remote blocks.

Commit:
throw exception when remote blocks size=0

SparkQA commented:
**[Test build #90280 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90280/testReport)** for PR 21212 at commit [`f9f0050`](https://github.com/apache/spark/commit/f9f005089bb930484298bf77b4721eef917447fb).

SparkQA commented:
**[Test build #90280 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90280/testReport)** for PR 21212 at commit [`f9f0050`](https://github.com/apache/spark/commit/f9f005089bb930484298bf77b4721eef917447fb).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90280/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2966/
Test PASSed.

SparkQA commented:
**[Test build #90283 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90283/testReport)** for PR 21212 at commit [`094d829`](https://github.com/apache/spark/commit/094d829ff5c36cdc358f4d6e635297ca4f4e400f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2968/
Test PASSed.

SparkQA commented:
**[Test build #90283 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90283/testReport)** for PR 21212 at commit [`094d829`](https://github.com/apache/spark/commit/094d829ff5c36cdc358f4d6e635297ca4f4e400f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90283/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

jinxing64 commented:
Thanks for merging !

squito commented:
@jinxing64 
> I guess your concern is ArrayBuffer will do lots of copy as size of elements grows, and we don't need fast random access in ShuffleBlockFetcherIterator

my concern wasn't the copy, my concern was the wasted space at the end of the ArrayBuffer, as that could waste a lot of space. ArrayBuffers are initialized with a size of 16: https://github.com/scala/scala/blob/v2.12.6/src/library/scala/collection/mutable/ArrayBuffer.scala#L63

and they double in size when they grow.  Both of those are good in general, but what if each of those ArrayBuffers only had one item in it?  Then you'd have an array with 15 empty slots.  Similarly if you had 17 elements, or 33 elements, etc.  I thought the memory inspection should tell you how many trailing nulls you had in your arrays.
